db/hints/manager.cc

```diff
@@ -446,6 +448,62 @@ locator::host_id manager::convert_ep_to_host_id(endpoint_id ep) const {
     return tmptr->get_host_id(ep);
 }
 
+future<> manager::rename_endpoint_directories() {
+    // We want to have a consistent topology throughout the calls in the lambda below.
+    // It's imperative because we cannot allow for a host ID to have mulitple IPs while
+    // this function is being executed.
+    //
+    // Imagine the following scenario of consecutive steps to visualize it:
+    //  (1) We rename the directory corresponding to ip_1 to host_id_1.
+    //  (2) The node corresponding to host_id_1 changes its IP to ip_2.
+    //  (3) We move on to processing a directory correspondiong to ip_2.
+    //      At this step, a failure occurs because ::rename(2) (so `seastar::rename_file` by extension too)
+    //      assumes that the new directory name (or more specifically, its path) doesn't exist.
+    //
+    // We utilize the useful property of token metadata here that ensures the topology stays immutable
+    // as long as we access it via this pointer.
+    const auto tmptr = _proxy_anchor->get_token_metadata_ptr();
+
+    // When upgrading Scylla, rename directories so that all of them use host IDs.
+    // If there is no mapping from an IP to a host ID, drop a directory by effectively removing it
+    // and its contents. Hinted Handoff is a "best-effort" mechanism -- there is no reason to deal
+    // with such edge cases, especially because this can only happen once -- during an upgrade.
+    co_await lister::scan_dir(_hints_dir, lister::dir_entry_types::of<directory_entry_type::directory>(),
+            [tmptr] (fs::path datadir, directory_entry de) -> future<> {
+        try {
+            locator::host_id_or_endpoint ep_or_hid{de.name};
+
+            if (!ep_or_hid.has_host_id()) {
+                // This either finds the corresponding host ID or throws an exception.
+                ep_or_hid.resolve(*tmptr.get());
+
+                // Old path: `<...>/<shard id>/<IP>`.
+                const fs::path ep_path = datadir / de.name;
+                // New path: `<...>/<shard id>/<host ID>`.
+                const fs::path hid_path = datadir / ep_or_hid.id.to_sstring();
+
+                // Here, we rely on the atomicity of ::rename(2) and the fact that we're the only
+                // ones performing actions related to `ep_path` and `hid_path` at the moment.
+                // See the System Calls Manual and this post:
+                //  https://stackoverflow.com/questions/7054844/is-rename-atomic
+                // for more information.
+                return io_check(rename_file, ep_path.c_str(), hid_path.c_str());
+            }
+
+            return make_ready_future<>();
```

Comment:
you can use a coroutine lambda and call `co_await io_check` above and `co_await lister::rmdir` below.